### DISCUSSION TITLE
How to check for undefined members of object on legacy browsers?

### DISCUSSION THREAD

**Discussion by mxdvl** [guardian/dotcom-rendering#9097] [RFCs]
As raised by @bryophyta in https://github.com/guardian/dotcom-rendering/pull/5849#issuecomment-1226980493, there are several ways of checking whether an method exists as a member of an object. We also have an auto-fixing ESLint rule that prevents us from using optional chaining on non-nullish values.

For example, the `window.performance.getEntriesByType` method is defined in [`lib.dom.d.ts`](https://github.com/microsoft/TypeScript/blob/c4eb37c8a0d5915d64e3992a91a34724aab19fe6/lib/lib.dom.d.ts#L11076), so we need to circumvent this rule to prevent errors in production.

Which options do people prefer?

```ts
// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- not all browsers have this method
const option1  = window.performance.getEntriesByType?.('measure') ?? [];

const option2  = 'getEntriesByType' in window.performance ? window.performance.getEntriesByType('measure') : [];

const option3  = typeof window.performance.getEntriesByType === 'function' in  ? window.performance.getEntriesByType('measure') : [];
```

**Comment by bryophyta** [comment]
Thanks for this! If `option1` is functionally equivalent to the others then I'd be inclined towards that? It looks like the issue here is essentially in the type definitions themselves, right? (I'm assuming that TS is basing its definitions here on a set of browsers that it's decided to support?) So being explicit that this is the source of the issue here seems like a good idea to me. But I don't have a super strong opinion on this :)
